### ISSUE TITLE
Skipping serialization of null values depending on the type

### ISSUE THREAD

**Issue by ampeixoto** [square/moshi#1289]
Hello!

I started working with moshi a couple of weeks ago, so maybe I am missing something trivial, but I spent already quite a bit of time trying to fix this without success, so here is my question.

Having the following reproducible code:
```
fun main() {

    val moshi = Moshi.Builder().add(OptionalAdapter).build()
    val objectToSerialize = DummyObject()
    val json = moshi.adapter(DummyObject::class.java).serializeNulls().toJson(objectToSerialize)
    println(json)

}

@JsonClass(generateAdapter = true)
data class DummyObject(val value: Int=123, val someNullable: String? = null,
                       val someNotPresent: Optional<String> = Optional.NotPresent,
                       val somePresent: Optional<String> = Optional.Present("aaaa"))


class OptionalAdapter<T>(private val valueAdapter: JsonAdapter<T>) : JsonAdapter<Optional<T>>() {
    @Suppress("UNCHECKED_CAST")
    override fun fromJson(reader: JsonReader) = Optional.Present(valueAdapter.fromJson(reader) as T)

    override fun toJson(writer: JsonWriter, value: Optional<T>?) {
        when (value) {
            is Optional.NotPresent -> writer.nullValue()
            is Optional.Present -> valueAdapter.serializeNulls().toJson(writer, value.value)
        }
    }

    companion object Factory : JsonAdapter.Factory {
        override fun create(type: Type, annotations: Set<out Annotation>, moshi: Moshi): JsonAdapter<*>? {
            return if (Types.getRawType(type) == Optional::class.java && annotations.isEmpty()) {
                val valueType = if(type is ParameterizedType) {
                    type.actualTypeArguments.get(0)
                } else {
                    //Should not happen
                    throw IllegalArgumentException()
                }
                return OptionalAdapter(moshi.adapter<Any>(valueType).nullSafe())
            } else {
                null
            }
        }
    }
}

sealed class Optional<out T> {
    val provided get() = this !is NotPresent
    abstract val value: T

    object NotPresent : Optional<Nothing>() {
        // have the IDE raise an error if the user knows a type is missing but still tries to access a value
        @Deprecated(
            "Cannot access a missing value",
            level = DeprecationLevel.ERROR,
            replaceWith = ReplaceWith("TODO(\"value is missing\")")
        )
        override val value: Nothing
            get() = error("cannot access provided field")
    }

    data class Present<out T>(override val value: T) : Optional<T>()
}
```

I would like to serialize as `{"value":123,"someNullable":null,"somePresent":"aaaa"}` instead of `{"value":123,"someNullable":null,"someNotPresent":null,"somePresent":"aaaa"}`, which is what is doing now.

Basically, I want to skip the serialization in case the type is `Optional.NotPresent`. Any suggestion?

**Comment by ampeixoto** [comment]
Found a solution:
```
override fun toJson(writer: JsonWriter, value: Optional<T>?) {
        when (value) {
            is Optional.NotPresent -> {
                val wasSerializeNulls = writer.serializeNulls
                writer.serializeNulls = false
                try {
                    writer.nullValue()
                } finally {
                    writer.serializeNulls = wasSerializeNulls
                }
            }
            is Optional.Present -> valueAdapter.serializeNulls().toJson(writer, value.value)
        }
    }
```
